crypto/s2n_ecdsa.c

```diff
@@ -32,8 +35,27 @@
 #include "crypto/s2n_openssl.h"
 #include "crypto/s2n_pkey.h"
 
+
 #define S2N_ECDSA_TYPE 0
 
+EC_KEY *s2n_unsafe_ecdsa_get_non_const(const struct s2n_ecdsa_key *ecdsa_key) {
+    if(ecdsa_key == NULL) {
+        return NULL;
+    }
+
+    /* pragma gcc diagnostic was added in gcc 4.6 */
+#if defined(__clang__) || S2N_GCC_VERSION_AT_LEAST(4,6,0)
```

Comment:
This should ideally be a compiler feature probe rather than a version check.

Reply:
Not 100% sure what this involves but I wasnt planning to address this in this PR since its already a bit complicated and we already do something similar [elsewhere](https://github.com/aws/s2n-tls/blob/6cf41a5ca382356c02d8a0825629154d8f5527d5/tls/s2n_connection.c#L526).

Even if we do a feature probe for clang/gcc wouldn't we still need some statement here. At that point I dont see the benefit.

Reply:
The benefit is we just check to see if the compiler we're using supports this pragma and if it does then we define `SUPPORTS_GCC_DIAGNOSTIC`. That way we don't have to encode information about compiler versions in our code.

Reply:
I'm fine with this being addressed separately, though, especially if there's similar patterns elsewhere.

Reply:
Every time we add to this list my heart gets heavier: #3327

Reply:
Agreed. Let's just open an issue or add to maddeleine's sad list for now.

Reply:
added to the existing list https://github.com/aws/s2n-tls/issues/3327